### PR TITLE
Bugfix 1853: Fetch error handling console

### PR DESCRIPTION
**Pull request by imranariffin**
### Pull Request checklist

- [ ] **Documentation**: If this feature has or requires documentation, the relevant docs have been updated.
- [x] **Changelog**: This PR updates the [changelog](https://github.com/iodide-project/iodide/blob/master/CHANGELOG.md) with any user-visible changes.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not

### Summary

Currently in master when executing fetch cell with this command
```
%% fetch
js: js: https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.11/lodash.min.js
```
the console will report the following
```
ERROR: js
    Error: 500 Internal Server Error (files/js)
```
which is not very helpful to the user, as described in #1853.

So this PR should resolve the above issue plus:
1. Report errors in the console with more helpful message on invalid fetch
2. Refactor the error handling code during parsing of fetch line such that errors are found out and reported early

### Sample changes in error outputs
Below are sample current vs new error reporting from different invalid fetch lines:

#### Fetch type appear more than once (original issue)
```
%% fetch
js: js: https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.11/lodash.min.js
``` 
Current:
```
ERROR: js
    Error: 500 Internal Server Error (files/js)
```
This PR:
```
Syntax error, invalid fetch url in:
    "js: js: https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.11/lodash.min.js"
```

#### Script fetch (css/js) provided with variable name when it shouldn't
```
%% fetch
js: varname = https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.15/lodash.core.js
```
Current:
```
ERROR: varname = https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.15/lodash.core.js
    Error: 404 Not Found (files/varname = https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.15/lodash.core.js)
```
This PR:
```
Syntax error, invalid fetch url in:
    "js: varname = https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.15/lodash.core.js"
```

### CODE REVIEW

File: src/editor/actions/fetch-cell-parser.js [hunk 1]
```diff
@@ -49,7 +46,61 @@ export function emptyLine(line) {
...
+    For script fetch (css & js), valid fetch content is:
+      "<fetch-url>"
+    For data fetch (arrayBuffer, blob, json & text), valid fetch content is:
+      "<variable-name> = <fetch-url>"
+  */
+  const fetchType = line.split(": ")[0];
+  const fetchContent = line.replace(new RegExp(`^${fetchType}: `), "").trim();
+
+  let fetchUrl;
+  if (fetchType.match(/^(js|css)$/)) {
```
Comment (wlach) [inline src/editor/actions/fetch-cell-parser.js]: Similar to the above, could we specify an array called `URL_FETCH_TYPES` and do:

```suggestion
  if (URL_FETCH_TYPES.includes(fetchType)) {
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@
 - Allow running simple-serve from a port other than 8000 (fixes #2444)
 - Allow blank content on notebook revision creation (fixes #2445) (#2455)
 - Enable user to clear console history with a click (#1015)
+- Report more specific errors in console when executing fetch cell
```
Comment (wlach) [inline CHANGELOG.md]:
```suggestion
- Report more specific errors in console when executing fetch cell (fixes #1853) (#2475)
```

File: src/editor/actions/fetch-cell-parser.js [hunk 3]
```diff
@@ -49,7 +46,61 @@ export function emptyLine(line) {
...
 }
 
-export function parseFetchCellLine(line) {
+export function missingFetchType(line) {
+  return !line.trim().match(/^\w+\s*: /);
+}
+
+export function validFetchType(line) {
+  const fetchType = line.trim().split(": ")[0];
+  return fetchType.trimLeft().match(/^(css|js|arrayBuffer|blob|json|text)$/);
```
Comment (wlach) [inline src/editor/actions/fetch-cell-parser.js]: Could we create an array called `VALID_FETCH_TYPES` in this file, and then do something like:

```suggestion
  return VALID_FETCH_TYPES.includes(fetchType.trimLeft());
```

**Comment by imranariffin** [comment]
> Thank you @imranariffin! this looks great to me-- I have a few very small requests but other than that it looks really good (and works well in my testing). I'm adding @bcolloran to the reviewers he has time to look at it, but I think I'm generally ok with landing this after my nits are addressed.

Thanks for the review! I have pushed some commits that should resolve the change requests. I also noticed that I forgot to include `plugin` as a valid fetch type so I have added it to the newly defined `VALID_FETCH_TYPE` constant.

**Comment by wlach** [comment]
Looks like @bcolloran is out, I'm going to go ahead and merge this one. Thank you @imranariffin! Let me know if you need some suggestions on what to work on next.

**Comment by imranariffin** [comment]
> Looks like @bcolloran is out, I'm going to go ahead and merge this one. Thank you @imranariffin! Let me know if you need some suggestions on what to work on next.

Thank you! Would love to hear your suggestion on issues for me to pick.

**Comment by wlach** [comment]
> > Looks like @bcolloran is out, I'm going to go ahead and merge this one. Thank you @imranariffin! Let me know if you need some suggestions on what to work on next.
> 
> Thank you! I'll be browsing for issues myself but would definitely love to hear your suggestion on issues for me to pick.

#1786 would be a logical followup to this one-- it's fetch related but has some other interesting angles to it.

**Comment by imranariffin** [comment]
@wlach Looks like a good one, can you assign it to me?

**Comment by wlach** [comment]
> @wlach Looks like a good one, can you assign it to me?

I'm having some issues assigning issues to you. I added you as a repo collaborator, which might help a bit. In any case, feel free to start working on it 😄

**Comment by imranariffin** [comment]
I see, that's weird. Anyways I have accepted the invite, thank you!

**Pull request merged**